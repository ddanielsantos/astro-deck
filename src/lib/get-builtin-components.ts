const BUILTIN_COMPONENTS = ["Head", "Header", "Notes", "StepList"];

export function getBuiltinComponents(slide: string) {
	const JSX_COMPONENT_MATCHER = /(?<=<)([A-Z][A-z]*)(?=[^/]*\/?>)/g;

	let components: RegExpMatchArray | string[] | null = slide.match(
		JSX_COMPONENT_MATCHER,
	);

	if (!components) {
		return [];
	}

	components = components.filter((c) => BUILTIN_COMPONENTS.includes(c));

	return [...new Set(components)];
}
