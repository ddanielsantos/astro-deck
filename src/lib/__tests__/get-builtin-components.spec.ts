import { it, expect } from "vitest";

import { getBuiltinComponents } from "../get-builtin-components.js";

it("should handle duplicated components", () => {
	const slide = `
        <Head>
            <title>astro-deck</title>
        </Head>

        <Header>

        header 1

        </Header>

        <Header>

        header 2

        </Header>

        <StepList >

            Uppercase text
            steplist    

        </StepList>

        <StepList/>
  `;

	const components = getBuiltinComponents(slide);

	expect(components).toMatchSnapshot();
	expect(components.length).toBe(3);
	expect(components).toStrictEqual(["Head", "Header", "StepList"]);
});

it("should ignore custom components", () => {
	const slide = `
        <Head>
            <title>astro-deck</title>
        </Head>

        <MyCustomComponent>
            my awesome title
        </MyCustomComponent>

        <Header>

        header 1

        </Header>

        <Header>

        header 2

        </Header>

        <StepList >

            Uppercase text
            steplist    

        </StepList>

        <StepList/>
  `;

	const components = getBuiltinComponents(slide);

	expect(components).toMatchSnapshot();
	expect(components.length).toBe(3);
	expect(components).toStrictEqual(["Head", "Header", "StepList"]);
});
