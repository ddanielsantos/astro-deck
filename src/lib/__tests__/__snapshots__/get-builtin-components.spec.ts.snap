// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should handle duplicated components 1`] = `
[
  "Head",
  "Header",
  "StepList",
]
`;

exports[`should ignore custom components 1`] = `
[
  "Head",
  "Header",
  "StepList",
]
`;
